tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/mapper/CollationMapper.java

```diff
@@ -68,88 +71,76 @@
-            // TODO: Fix CLDR (https://unicode-org.atlassian.net/projects/CLDR/issues/CLDR-13131)
-            if (icuData.getName().equals("root")) {
-                icuData.replace(RB_STANDARD_SEQUENCE, "");
-                // TODO: Collation versioning probably needs to be improved.
-                icuData.replace(RB_STANDARD_VERSION, cldrVersion);
-            }
+    private final IcuData icuData;
+    private final String cldrVersion;
+
+    private CollationMapper(IcuData icuData, String cldrVersion) {
+        this.icuData = checkNotNull(icuData);
+        this.cldrVersion = checkNotNull(cldrVersion);
+        // Super special hack case because the XML data is a bit broken for the root collation
+        // data (there's an empty <collation> element that's a non-leaf element and thus not
+        // visited, but we should add an empty sequence to the output data.
+        // TODO: Fix CLDR (https://unicode-org.atlassian.net/projects/CLDR/issues/CLDR-13131)
+        if (icuData.getName().equals("root")) {
+            icuData.replace(RB_STANDARD_SEQUENCE, "");
+            // TODO: Collation versioning probably needs to be improved.
+            icuData.replace(RB_STANDARD_VERSION, cldrVersion);
         }
+    }
 
-        @Override
-        public void visitPrefixStart(CldrPath prefix, Context ctx) {
-            if (COLLATIONS.matchesPrefixOf(prefix)) {
-                ctx.install(this::collectRules);
-            } else if (SPECIAL.matchesPrefixOf(prefix)) {
-                ctx.install(this::maybeAddSpecial);
-            }
+    private void collectRule(CldrValue v) {
+        String type = COLLATION_TYPE.valueFrom(v);
+        RbPath rbPath = RbPath.of("collations", type, "Sequence");
+
+        // WARNING: This is almost certainly a bug, since while @type can have the value
+        // "short" it can also have other values. This code was copied from CollationMapper
+        // which has the line;
+        //   isShort = attr.getValue("alt") != null;
+        // TODO: Raise a ticket to examine this.
+        boolean isShort = COLLATION_RULE_ALT.optionalValueFrom(v).isPresent();
+
+        // Note that it's not clear why there's a check for "contains()" here. The code
+        // from which this was derived is largely undocumented and this check could have
+        // been overly defensive (perhaps a duplicate key should be an error?).
+        if (isShort || !icuData.getPaths().contains(rbPath)) {
+            RbValue rules = RbValue.of(
+                LINE_SPLITTER.splitToList(v.getValue()).stream()
+                    .map(CollationMapper::removeComment)
+                    .filter(s -> !s.isEmpty())::iterator);
+            icuData.replace(rbPath, rules);
+            icuData.replace(RbPath.of("collations", type, "Version"), cldrVersion);
         }
+    }
 
-        private void collectRules(CldrValue v) {
-            CldrPath p = v.getPath();
-            if (COLLATION_RULE.matchesSuffixOf(p)) {
```

Comment:
What used to be conditional code with explicit matching logic now becomes 2 different callbacks in the new code.